Add options for constructing an ORANGE geometry

elliottbiondo opened the pull request:
This MR adds a `ConstructionOptions` struct in order to facilitate the study of performance-sensitive parameters that govern ORANGE geometry construction.

elliottbiondo committed:
Change BIH::Input to BIH::Options

elliottbiondo committed:
Construct BIH based options in OrangeInput

elliottbiondo commented:
@sethrj can you take a preliminary look at this? We already have  `g4org::Options` -> `InputBuilder::Options`, so the addition of `OrangeInput::ConstructionOptions` feels suspicious.

sethrj commented on src/orange/OrangeInput.hh:
```diff
@@ -159,6 +161,19 @@ struct RectArrayInput
+struct ConstructionOptions
+{
+    //! Options for Bounding Interval Hierarchy (BIH) construction
+    detail::BIHBuilder::Options bih_options;
```
It might make it more messy in the short term, but I think the preferred pattern will look more like the newer celeritas code: `inp/Foo.hh` has the setup options for `Foo`. So we'd have `orange/inp/Bih.hh` which would have `struct inp::Bih` and that would replace `BihBuilder::Options`. We'd likewise move `orange/gorg/Options.hh` to `orange/inp/Converter.hh`. Finally we'd also move `OrangeInput` to `orange/inp/Model.hh` or similar.

elliottbiondo replied:
FWIW, I think `Options` makes more sense, lest we overload the word "input"; we already have an `orangeinp` folder. Likewise, below, I think semantically the input to the `BIHBuilder` are the bounding boxes, whereas the `min_split_size` is an option specifying how the process is done.

elliottbiondo replied:
The `orange/inp` folder doesn't currently exist. I can make, but I am still not clear on how InputBuilder::Options fits in (seems like that should be defined at the `OrangeInput` level).

I also, I think `g4org::Options` should consist of: 1) a high-level struct will all general OrangeOptions, that is also used for a non-G4 workflow to specify all ORANGE options in one place, and 2) additional G4-specific options.

sethrj replied:
The weirdness/confusion is because we've been inconsistent in the past and are transitioning toward having an `inp` namespace that has *all* of the input structs that were formerly called Input/Options. This namespace is to be the primary interface with the outside parts of the code: see #2029.

In a hypothetical/future refactorization that I am talking about:
- `celeritas::OrangeInput` -> `celeritas::inp::Orange`
- `celeritas::detail::BIHBuilder::Options` -> `celeritas::inp::BihBuilder`
- `celeritas::g4org::Options` -> `celeritas::inp::Converter` (?)
- `celeritas::orangeinp` -> `celeritas::setup` (?)
- `celeritas::g4org` -> `celeritas::setup::g4org` (?)

I don't know what's the right path between that future and our present, but I would at least like:
- BIH builder doesn't include half of orange transitively via the OrangeInput (since VariantSurfaces)
- OrangeInput doesn't include and rely on a detail namespace
- Mlutiple layers of nested options with only one member per struct

elliottbiondo replied:
Not sure what you mean by these two:

> BIH builder doesn't include half of orange transitively via the OrangeInput (since VariantSurfaces)

> Mlutiple layers of nested options with only one member per struct

sethrj replied:
- If `BIHBuilder` included `OrangeInput.hh` then it would end up pulling in a bunch of "setup-time" files. I guess it doesn't really matter since those are just included by the other setup code: when I wrote that I was thinking it was *runtime* (device) code that included it.
- I think I meant *avoiding* the second one? e.g.: avoid `struct BarOpts { int value; }; struct FooOpts { BarOpts bar_opts; }`

elliottbiondo replied:
That all makes sense. I was confused because BIHBuilder did not include OrangeInput.hh, but you were speaking hypothetically. And yes, at some point maybe an options struct in `inp/orange.hh` can eliminate the `inp/Bih.hh` file.

sethrj commented on src/orange/detail/BIHBuilder.hh:
```diff
@@ -51,11 +51,14 @@ class BIHBuilder
-    //! Input parameters
-    struct Input
+    //! Options parameters
+    struct Options
```
if we move the struct, this'll become `using Input = inp::BihBuilder;` as is done in the other update celeritas classes.

elliottbiondo committed:
Use inp::BihBuilder struct

elliottbiondo commented:
@sethrj I am finally getting around to fixing this MR since I want to do the performance tests this week.

I moved the options struct to `inp/` and added a `CELER_BIH_MIN_SPLIT_SIZE` environment variable. This is ready for another look.

sethrj commented:
@elliottbiondo Thanks for picking this back up! Just mark it as "ready" to ask me for another review :)

sethrj commented on doc/usage/execution/environment.rst:
```diff
@@ -18,35 +18,36 @@ has a default value but can be overridden with any boolean value.
+ ======================== ========= ==========================================
+ Variable                 Component Brief description
+ ======================== ========= ==========================================
+ CELER_BIH_MIN_SPLIT_SIZE orange    Set BIH ``min_split_size``
```
Rename `ORANGE_`... and since this is a temporary hack (we're going to add more options via a cleaner interface) don't bother putting it in the environment docs. No one will know what it means...

sethrj commented on src/orange/detail/UnitInserter.hh:
```diff
@@ -39,7 +39,9 @@ class UnitInserter
-    UnitInserter(UniverseInserter* insert_universe, Data* orange_data);
+    UnitInserter(UniverseInserter* insert_universe,
+                 Data* orange_data,
+                 ConstructionOptions* opts);
```
Why are the options passed by mutable pointer?

sethrj commented on src/orange/inp/Bih.hh:
```diff
@@ -0,0 +1,28 @@
+struct BIHBuilder
+{
+    //! Minimum number of bboxes needed to trigger a partitioning attempt
+    size_type min_split_size = 2;
```
Missing include `corecel/Types.hh`

sethrj commented on src/orange/orangeinp/InputBuilder.cc:
```diff
@@ -13,6 +13,7 @@
 #include "corecel/io/JsonPimpl.hh"
 #include "corecel/io/Logger.hh"
 #include "corecel/io/ScopedTimeLog.hh"
+#include "corecel/sys/Environment.cc"
```
```suggestion
#include "corecel/sys/Environment.hh"
```

sethrj reviewed (approved):
Needs a couple of fixes but I think this is the right way to do it for the short term. Thanks!

elliottbiondo committed:
Change CELER_BIH_MIN_SPLIT_SIZE to ORANGE_BIH_MIN_SPLIT_SIZE

elliottbiondo committed:
Pass ConstructionOptions by const pointer

elliottbiondo committed:
Add missing include

elliottbiondo committed:
Update src/orange/orangeinp/InputBuilder.cc

FIx typo: .cc -> .hh

Co-authored-by: Seth R. Johnson <johnsonsr@ornl.gov>

elliottbiondo committed:
Merge branch 'options' of https://github.com/elliottbiondo/celeritas into options

elliottbiondo commented:
@sethrj RFAL

sethrj commented:
Note that we *very rarely* use const pointers; only as an "optional reference". Use const references in most cases. The reason for mutable pointers was following an older version of the Google style guide, which points out that passing by reference "hides" whether the argument is going to be modified or not. (This is I think one big shortcomings of C++ : sometimes you want signatures to be identical between mutable/immutable, but passing mutable arguments makes it hard to see from the caller's point of view what data is being read vs written.)

elliottbiondo committed:
Merge branch 'develop' into options
